### ISSUE TITLE
Asciidoc JavaFx WebView hangs/crashes IntelliJ

### ISSUE THREAD

**Issue by craftey** [asciidoctor/asciidoctor-intellij-plugin#133]
AsciiDoc Plugin Version: 0.14.1 / 014.2 / 0.15.0
IDE Name and Version: 2016.2.1
OSX: 10.11.6

When opening an asciidoc file (adoc) with 
```
Settings --> AsciiDoc: 
Preview Browser = JavaFx WebView 
Default editor layout = Show editor and Preview 
```
then IntelliJ gets unresponsive (spinning ball). I can only force kill IntelliJ.

Setting 
`Preview Browser = Swing `
fixes the issue. Opening an adoc now does not crash. Even going now to settings and switch back to 
`Preview Browser = JavaFx WebView `
successfully changes the preview of the open adoc to the new JavaFx style. I can even CMD-W the editor-view and reopen the adoc again. 

But after a while, maybe restarting IntelliJ, or after working a bit, maybe a system hibernate over night, comming back to my adoc crashes IntelliJ as described before. 

Some Tests with setting 
`Preview Browser = JavaFx WebView `
enabled: 
1) with open adoc, quit intellij, open intellij, watch opened adoc --> **no-crash** 
2) Close the adoc, quit intellij, open intellij, compile and run a project, open the adoc with CMD-Shift-O --> **crash**. 
3) 2 times tested: Close the adoc, quit intellij, open intellij, open the adoc with CMD-E --> **no-crash** o_O , Cmd-W, CMD-Shift-O --> **crash**. 
4) Close the adoc, quit intellij, open intellij, open the adoc with CMD-Shift-O --> **crash**. 
5) 2 times tested: Close the adoc, quit intellij, open intellij, open the adoc with CMD-E --> **no crash**, CMD-W, CMD-E --> **no crash**, CMD-W, CMD-E --> **no crash**, CMD-W, CMD-E --> **no crash**, ..., CMD-W, CMD-E --> **crash**. Finally it crashes.

CMD-Shift-O opens the any file search 
CMD-E opens the recent file lists
CMD-W closes the editor tab of current file

Setting 
`Preview Browser = Swing `
fixes the issue again.

Also tested JavaFx WebView with Markdown-files when Plugin "Markdown support" is installed: I could not produce any crash via Test 5).

Cheers, Craftey

**Comment by craftey** [comment]
I send some logs via mail to Erik Pragt and Alexander Schwartz.

**Comment by ahus1** [comment]
Thank you for the detailed explanation. I'll have a look at the log file.

* Is there anything special to know about the adoc file (graphics of different kinds, etc.?)

Thanks, Alexander

**Comment by craftey** [comment]
The adoc files I use for the tests just contain headlines, text and code blocks.
Since Test 5) was quite reproducible, can you reproduce the issue with Test 5?

```
Settings --> AsciiDoc: 
Preview Browser = JavaFx WebView 
Default editor layout = Show editor and Preview 

TEST 5: 
Open adoc, close the adoc, quit Intellij, open Intellij, 
Open the adoc with CMD-E --> no crash, 
CMD-W, CMD-E --> no crash, 
CMD-W, CMD-E --> no crash, 
CMD-W, CMD-E --> no crash, 
..., 
CMD-W, CMD-E --> crash. Finally it crashes.

CMD-E opens the recent file lists
CMD-W closes the editor tab of current file
```
Note: I experience a hang (on OSX a spinning mouse wheel, maybe indication a gui callback waiting for something), not a crash. When IntelliJ hangs, cpu load starts at 10% und step by step climbs up to 25%. Then I took the only action possible, force quit (kill -9).

My `~/java_error_in_idea.hprof` file is an old one (July 13th), not related to the asciidoc hangs, which I reproduced right now again.

I also tested v0.14.2 and v0.15.0. Both version reproduce the hang with Test 5.

**Comment by voidmain** [comment]
Same issue here. I can send in additional logs if you need them.

**Comment by ahus1** [comment]
I tracked down a memory leak in JavaFX WebView. Images are cached in their uncompressed for the complete lifecycle of the JVM. Search the web for WCImageDecoderImpl and com.sun.javafx.iio.ImageFrame. I didn't find a proper workaround for it.

The animated image that shows a screen cast that is referenced from the README.adoc of this project expands to about 300 MB RAM usage for the lifetime of the JVM. Of course smaller and non-animated images will result in smaller memory footprint (like 10-20 K RAM).

I also found that there is often no heap dump (although IntelliJ requests it). Instead the IDE writes in the home directory a java_error_idea_xxxx.log that captures minimal information. 

You'll experience this problem if you reference big images your docs. 

The v14.0.2 and v0.15.0 are also problematic as they load images local the the AsciiDoc document on every rendering to allow for changing images. The normal behavior of a WebView is that it will show only the first version of an image and no changed. 

I'll will improve this in the next release to save some memory by only reloading images when they have changed. A future version might also have a switch "don't show images to save JVM memory" :-(

**Comment by craftey** [comment]
@ahus1 Could you reproduce the issue with Test 5 with an adoc without images?
Besides, the same test with plugin "Markdown Support" and a markdown file does not lead not to a hang of IntelliJ.

**Comment by ahus1** [comment]
I've been able to reproduce crashes - some with out of memory, sometime the JVM just crashes on Windows when using JavaFX. Im Currently getting better results when switching to 64bit Oracle JDK - https://intellij-support.jetbrains.com/hc/en-us/articles/206544879-Selecting-the-JDK-version-the-IDE-will-run-under. 

I'll see how this goes. I'm thinking about limiting JavaFX to 64bit JVMs.

**Comment by ahus1** [comment]
A new preview release 0.15.1 is available that restricts JavaFX preview to 64bit. I've also backported this to 0.14.3 that is available on Jetbrains plugin repository. 

This works quite stable for me even for big documents with IntellJ IDEA 2016.2 on Win 7 & Oracle JDK 64bit. On Win 7 & OpenJDK 32 (the included JDK) it proved quite unstable for me.

This should work on Linux, Mac and Windows as long as you are using 64bit JDK. 

> If you want to use JavaFX preview, you'll need to run IntelliJ with 64bit Java. For IntelliJ you'll need to have an environment variable `IDEA_JDK` pointing to the installation folder of a 64bit Java. See https://intellij-support.jetbrains.com/hc/en-us/articles/206544879-Selecting-the-JDK-version-the-IDE-will-run-under for more information.

**Comment by craftey** [comment]
Now with `0.14.3` or `0.15.1` I cannot select JavaFx in the asciidoc settings anymore. I configured IntelliJ to use an external JVM: I checked the about-info-panel of IntellIJ to confirm that `JRE: 1.8.0_102-b14 x86_64` is uses.
Tested on OSX 10.9.5 this time.

**Comment by craftey** [comment]
@ahus1, I just wanted to point again to the observation that the plugin "Markdown Support" also uses JavaFx WebView and does not show the behavior we are hunting here with asciidoc. The file I used lately for testing had only 5 lines of text. So maybe the cause of the issue is not the jdk and not the a memory leak in JavaFX WebView. Maybe it is something different. At least I come to this conclusion from my simple observations with asccidoc vs markdown. Dont hesitate to tell me if I am wrong.

**Comment by ahus1** [comment]
@craftey - I see that my 64bit detection for Mac is wrong. Expect a new preview version soon.

**Comment by ahus1** [comment]
@craftey - a new [preview release 0.15.2](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases/tag/0.15.2) is available. You should be able to enable JavaFX from the settings again. This tests for the string "x86_64" in the JVM architecture.

I hope this prevents the crashes you've experiencing. If not, please proceed as follows: 

* delete the exiting heap dump java_error_in_idea.hprof in you user's home directory. Otherwise IntelliJ will not create a new one on the next crash. 
* provide me with the java_error_in_idea_XXXX.log files. This will contain crash information from the JVM. **Please email them to me separately.** Please also send (if available) also the java_error_in_idea_XXXX.log files from the crashes you've experienced before you switched to the 64bit JDK.  
* Even more crash information might be available if you start IntelliJ from a command line window. Once the IDE crashes it will print information on the console.

Good luck!

**Comment by craftey** [comment]
I have tested version `0.15.2`. I now can select JavaFx. The issue is **still reproducible** with test 5. 
I also renamed `~/java_error_in_idea.hprof` to `~/java_error_in_idea.hprof.<date>`. And I started IntelliJ in Terminal via `/Applications/IntelliJ\ IDEA.app/Contents/MacOS/idea`.
When I reproduce the hang, nothing is appended to the console out log. And no new `~/java_error_in_idea.hprof` is generated.
IntelliJ does not respond, shows the spinning ball and I only can kill -9.

I have the feeling that with version `0.15.2` and external JVM, the issue is now faster reproducibel, often with the first try opening a simple adoc.

I also found out, that the issue is **not reproducible** on `Ubuntu 2016.4` on the machine of a colleague. We used the current repository version `0.14.3`and the integrated JVM for that test.

Conclusion: The issue might be related to OSX.

If possible please try to reproduce the hang on a mac of a friend/colleague or in a virtual machine with OSX. Thanks for any solution or steps towards it.

**Comment by minibiti** [comment]
Hi,
I have problem when using Ubuntu 14.04 but not when using Ubuntu 16.04.
I hope this can help find the problem :)

**Comment by ahus1** [comment]
Haven't come near an OSX here yet. All I experienced was crashes (not hangs) when using Windows 32bit Java. These crashes were gone after I switched to 64bit Java as stated earlier. 

I tried test 5 with several runs on Windows and 64 bit - no crashes, no hangs. 

I'll look into Ubuntu on a VM when I have more time. 

Until then it would be helpful if someone who is experiencing this could post a thread dump to identify the hang. On Linux/OSX this should work by "kill -3 PROCESSID".  IntelliJ also posts thread dumps to the user's home directory .IntelliJIdea2016.2\system\log - but I'm not sure if it does this in this situation as well.

**Comment by craftey** [comment]
I finally found the threadDump folder: ~/Library/Logs/IntelliJIdea2016.2
(thx to https://intellij-support.jetbrains.com/hc/en-us/articles/207241085)

About manually creating  thread dumps and IntelliJ' s auto detection of hangs: https://intellij-support.jetbrains.com/hc/en-us/articles/206544899-Getting-a-thread-dump-when-IDE-hangs-and-doesn-t-respond

For now I created thread dumps with the automatic hang detection of IntelliJ. I ll send them via mail.

**Comment by ahus1** [comment]
Hello @craftey - thank you for the detailed reports. I've had a look at the thread dump and spotted a clue. 

I've uploaded another preview version [0.15.4](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases/tag/0.15.4) - please download and install it. It should no longer wait on the CountDownLatch any more. 

If the problem persists, please send again some thread dumps.

**Comment by craftey** [comment]
I have tested `v0.15.4` on OSX 10.9.5 with `Test 5`.

`Result:`` `` `` ``It just work's`` `` `` ` :-)

I repeated the loop from `Test 5` about 10 times. The issue was not reproducible. I am convinced the issue is fixed now. Great that you could spot it. Thx a lot.

**Comment by ahus1** [comment]
Finally! Thank you for your patience. Too bad I didn't think of the stack trace earlier and focused on reproducing it on Windows. 

I learned a lot about about IntelliJ. I'll promote 0.15.4 to the IntelliJ repo now.

**Comment by mojavelinux** [comment]
:beers: tonight!

**Comment by minibiti** [comment]
Hi, I am sorry to report that it still crashed my IntelliJ 2016.2 running on Ubuntu x64 16.04

**Comment by ahus1** [comment]
@minibiti - please open a new ticket for this describing your situation: a document, your setup (IDE version, OS), and what you experience: it is a crash (application terminates) or a freeze (application still running, but no interaction possible)?

I'll also need the IntelliJ Log and the thread dump folder. Please see crafteys comment above where to find them. 

Please send the logs via email. You'll find my email address when you look into my Github profile.
